core/src/autogluon/core/models/ensemble/fold_fitting_strategy.py

```diff
@@ -0,0 +1,144 @@
+import copy
+import logging
+import os
+import time
+from abc import abstractmethod
+
+from ...utils.exceptions import TimeLimitExceeded
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractFoldFittingStrategy(object):
+
+    @abstractmethod
+    def schedule_fold_model_fit(self, model_base, fold_ctx, kwargs):
+        """
+        Schedule fold model training. By design this part is supposed to be 'lazy' evaluator no actual training is performed here.
+        Distributed fitters will handle jobs scheduling here.
+        """
+        pass
+
+    @abstractmethod
+    def after_all_folds_scheduled(self):
+        """
+        Method is called when all the folds are scheduled.
+        Local fitters will perform training here.
+        Distributed fitters will handle job handles and results retrieval here.
+        """
+        pass
+
+
+class SequentialLocalFoldFittingStrategy(AbstractFoldFittingStrategy):
+    """
+    This strategy fits the folds locally in a sequence.
+    """
+
+    def __init__(self, bagged_ensemble_model, X, y, sample_weight, time_limit, time_start, models, oof_pred_proba, oof_pred_model_repeats):
```

Comment:
Consider adding type hints

Reply:
Added except for `bagged_ensemble_model` - that creates circular dependencies in imports